prime-router/src/main/kotlin/history/db/ReportGraph.kt

```diff
@@ -0,0 +1,100 @@
+package gov.cdc.prime.router.history.db
+
+import gov.cdc.prime.router.azure.DatabaseAccess
+import gov.cdc.prime.router.azure.db.Tables.COVID_RESULT_METADATA
+import gov.cdc.prime.router.azure.db.Tables.REPORT_FILE
+import gov.cdc.prime.router.azure.db.Tables.REPORT_LINEAGE
+import gov.cdc.prime.router.azure.db.tables.pojos.CovidResultMetadata
+import gov.cdc.prime.router.common.BaseEngine
+import org.apache.logging.log4j.kotlin.Logging
+import org.jooq.CommonTableExpression
+import org.jooq.Record
+import org.jooq.Record2
+import org.jooq.impl.DSL
+import org.jooq.impl.DSL.selectDistinct
+import org.jooq.impl.SQLDataType
+import java.util.UUID
+
+class ReportGraph(
+    val db: DatabaseAccess = BaseEngine.databaseAccessSingleton
+) : Logging {
+
+    /**
+     * Returns all the metadata for the items in the past in reports; will recursively walk up the report lineage
+     * and then filter down to reports where the sender is set in order to find the metadata rows
+     *
+     * @param childReportIds the list of ids to start from
+     *
+     */
+    fun getMetadataForReports(childReportIds: List<UUID>): List<CovidResultMetadata> {
+        val lineage = lineageCte(childReportIds)
+
+        val sourceReportIds =
+            sourceReportsCte(lineage)
+
+        val metadata = metadataExpression(sourceReportIds)
+
+        return db.transactReturning { txn ->
+            DSL.using(txn)
+                .withRecursive(lineage)
+                .with(sourceReportIds)
+                .with(metadata)
+                .selectDistinct(metadata.asterisk())
+                .from(
+                    metadata,
+                ).fetchInto(CovidResultMetadata::class.java)
+        }
+    }
+
+    private fun metadataExpression(sourceReportIds: CommonTableExpression<Record>) =
```

Comment:
These are CTEs that can be combined in order to get different slices of data by combining the report lineage with report file and the covid metadata table